rust/connlib/tunnel/src/device_channel/tun_linux.rs

```diff
@@ -456,4 +530,110 @@ nameserver 100.100.111.2
 "
         );
     }
+
+    /// Returns `Ok(())` if the file at the given path contains the expected sentinels
+    ///
+    /// Return `Err(_)` if the file can't be parsed, the sentinels don't match, or
+    /// any DNS options are set in the file.
+    fn check_resolv_conf(path: &Path, expected_sentinels: &[IpAddr]) -> Result<()> {
+        let text = std::fs::read_to_string(path).context("could not read file")?;
+        let parsed = resolv_conf::Config::parse(text)?;
+
+        let mut expected = resolv_conf::Config::new();
+        expected_sentinels
+            .iter()
+            .for_each(|addr| expected.nameservers.push((*addr).into()));
+
+        ensure!(
+            parsed == expected,
+            "Parsed resolv config didn't match expected resolv config"
+        );
+        Ok(())
+    }
+
+    // Not shared with prod code because prod also writes the "Generated by" comment,
+    fn write_resolv_conf(path: &Path, nameservers: &[IpAddr]) -> Result<()> {
+        let mut conf = resolv_conf::Config::new();
+        conf.nameservers = nameservers.iter().map(|addr| (*addr).into()).collect();
+        std::fs::write(path, conf.to_string())?;
+        Ok(())
+    }
+
+    /// The original resolv.conf should be backed up, and the new one should only
+    /// contain our sentinels.
+    #[tokio::test]
+    async fn resolv_conf_happy_path() -> Result<()> {
+        // Try not to panic, it may leave temp files behind
+        // Using `TempDir` instead of `tempfile` because I need the path, and instead
+        // of `NamedTempFile` because those get deleted immediately on Linux, which confuses
+        // `configure_resolv_conf when it tries to read from the path.
+        let temp_dir = tempfile::TempDir::with_prefix("firezone-dns-test")?;
+
+        let resolv_path = temp_dir.path().join("resolv.conf");
+        let backup_path = temp_dir.path().join("resolv.conf.firezone-backup");
+
+        write_resolv_conf(&resolv_path, &[GOOGLE_DNS.into()])?;
+
+        configure_resolv_conf(
+            &[IpAddr::from([100, 100, 111, 1])],
+            &resolv_path,
+            &backup_path,
+        )
+        .await?;
+
+        check_resolv_conf(&resolv_path, &[IpAddr::from([100, 100, 111, 1])])
+            .context("{resolv_path}")?;
```

Comment:
`.context()` doesn't expand variables, you'll have to use:

```suggestion
            .with_context(|| format!("{resolv_path}"))?;
```